### PR TITLE
Refactor symbolics

### PR DESCRIPTION
**Pull request by MaximilianAlgehed**
This allows us to add other symbolics than `SymToken` easily.

### CODE REVIEW

**Review by sjoerdvisscher — APPROVED** [review]
Really nice use case for `barbie`!

File: quickcheck-contractmodel/src/Test/QuickCheck/ContractModel/Internal/Symbolics.hs [hunk 1]
```diff
@@ -1,20 +1,157 @@
...
+
+mappendSymIndexF :: forall f. (AllBF Semigroup f SymIndexF, Show (SymIndexF f))
+                 => (forall a. f a -> Set String)
+                 -> SymIndexF f
+                 -> SymIndexF f
+                 -> SymIndexF f
+mappendSymIndexF toSet s s'
+  | and (Set.disjoint <$> bmapConst toSet s
+                      <*> bmapConst toSet s') = bzipWithC @(ClassF Semigroup f) (<>) s s'
+  | otherwise = error $ unlines [ "Non-unique SymIndexF" ]
```
[the remark this answers is not in the document]
Comment (MaximilianAlgehed) [reply quickcheck-contractmodel/src/Test/QuickCheck/ContractModel/Internal/Symbolics.hs:62]: Whoops... That's evidence of some experimentation that's been lost...

I'll fix it by restoring the original implementation...

**Comment by MaximilianAlgehed** [comment]
> Really nice use case for `barbie`!

Right? Although! I'm not yet convinced that the barbie library is the right implementation of this pattern yet. The `Const` functor stuff is all a bit clunky because there are some missing instances that don't really have any right to be missing....

**Comment by sjoerdvisscher** [comment]
> > Really nice use case for `barbie`!
> 
> Right? Although! I'm not yet convinced that the barbie library is the right implementation of this pattern yet. The `Const` functor stuff is all a bit clunky because there are some missing instances that don't really have any right to be missing....

Is this just an omission from the `barbie` library? Or are these instances impossible to implement?

**Comment by MaximilianAlgehed** [comment]
I think they are impossible given the implementation of barbie. But I'm
also not sure how to do it better.

Den tors 20 apr. 2023 12:30Sjoerd Visscher ***@***.***> skrev:

**Pull request merged**